modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/NPMRegistryResolutionStateDigestUtil.java

```diff
@@ -0,0 +1,204 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.js.loader.modules.extender.internal.npm;
+
+import com.liferay.frontend.js.loader.modules.extender.internal.npm.dynamic.DynamicJSModule;
+import com.liferay.frontend.js.loader.modules.extender.npm.JSModule;
+import com.liferay.frontend.js.loader.modules.extender.npm.JSModuleAlias;
+import com.liferay.frontend.js.loader.modules.extender.npm.JSPackage;
+import com.liferay.frontend.js.loader.modules.extender.npm.JSPackageDependency;
+import com.liferay.frontend.js.loader.modules.extender.npm.NPMRegistry;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.patcher.PatcherUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.io.UnsupportedEncodingException;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Computes a digest (hash) with ALL values that can make module resolutions
+ * change.
+ *
+ * The data added to the digest is based on package names, versions,
+ * dependencies and any other information that, when changed, may alter the way
+ * modules are resolved.
```

Comment:
How does this work with local development where we might be changing an npm package locally? For example, if we used `yarn link` for a package, would the resolutions still be cached? Or is it just the path resolutions that are resolved?

Reply:
Mmm, I think it would fail depending on how much you change the package. So, if you don't change package name, version or deps, resolutions will be cached. Then, if the module for which resolutions were cached has changed any of its deps, the AMD loader won't get some of them and/or will retrieve others that are not needed.

However, we are already recommending disabling the cache for development, because there are more problems than this one due to the cache, so...

In any case, it's a good point. We need to be vigilant.